.github/workflows/tests.yml

```diff
@@ -0,0 +1,423 @@
+  # is to compile (at least at this writing).
+  build-zope_interface:
+    # Sigh. Note that the matrix must be kept in sync
+    # with `test`, and `docs` must use a subset.
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        python-version: [2.7, pypy-2.7, pypy-3.6, 3.5, 3.6, 3.7, 3.8, 3.9]
+        os: [ubuntu-20.04, macos-latest]
+        exclude:
+          - os: macos-latest
+            python-version: pypy-2.7
+          - os: macos-latest
+            python-version: pypy-3.6
+          - os: macos-latest
+            python-version: 3.5
+
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      ###
+      # Caching.
+      # This actually *restores* a cache and schedules a cleanup action
+      # to save the cache. So it must come before the thing we want to use
+      # the cache.
+      ###
+      - name: Get pip cache dir
+        id: pip-cache
+        run: |
+          echo "::set-output name=dir::$(pip cache dir)"
+
+      - name: pip cache
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.pip-cache.outputs.dir }}
+          key: ${{ runner.os }}-pip-${{ matrix.python-version }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+
+      - name: Install Build Dependencies
+        run: |
+          pip install -U pip
+          pip install -U setuptools wheel twine
+          pip install -U coveralls coverage
+
+      - name: Build zope.interface
+        run: |
+          # Next, build the wheel *in place*. This helps ccache, and also lets us cache the configure
+          # output (pip install uses a random temporary directory, making this difficult).
+          python setup.py build_ext -i
+          python setup.py bdist_wheel
+          # Also install it, so that we get dependencies in the (pip) cache.
+          pip install -U coverage
+          pip install -U 'faulthandler; python_version == "2.7" and platform_python_implementation == "CPython"'
+          pip install .[test]
+
+      - name: Check zope.interface build
+        run: |
+          ls -l dist
+          twine check dist/*
+      - name: Upload zope.interface wheel
+        uses: actions/upload-artifact@v2
+        with:
+          name: zope.interface-${{ runner.os }}-${{ matrix.python-version }}.whl
+          path: dist/*whl
+      - name: Publish package to PyPI (mac)
+        # We cannot 'uses: pypa/gh-action-pypi-publish@v1.4.1' because
+        # that's apparently a container action, and those don't run on
+        # the Mac.
+        if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags') && startsWith(runner.os, 'Mac')
+        env:
+          TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
+        run: |
+          twine upload --skip-existing dist/*
```

Comment:
Why do the upload before running tests?

Reply:
One basic reason, with a rationale.

Most importantly, making a release usually means using `zest.releaser`, which will upload the `.tar.gz` source file. Users that then attempt to just install `zope.interface` to get the latest version will have to have to be able to build C extensions while they wait for the new wheels to be built and uploaded, which can take awhile. I've had numerous issues opened (in other projects) about "hey I can't install your project" or "your newest release is missing wheels." Consequently, when a release is made, I try to optimize the build process to start getting wheels out to PyPI ASAP. People are both faster at noticing new releases, and also not pinning version requirements, than one might think.

I feel this is safe because at the time the release is made, CI was known to be green, and there are no code changes from the last CI run, only the textual things that `zest.releaser` does.

Reply:
I guess what I find strange is that only MacOS builders do the upload early; manylinux builds do it at the very end.  I suppose Linux users don't find the lack of binary wheels an inconvenience.

Reply:
For i686/x86_64, the *entire* set of manylinux wheels is built and uploaded in less than two minutes. These jobs are both kicked off immediately (have no dependencies) and run in parallel with the other build steps. They don't do testing either. So the wait is minimal.

Things are a bit different for the aarch64 wheels, which *do* test first, and the Windows wheels (which also test first, but also run separately in parallel to the process here). Refactoring those could be done, but I would consider that a separate task.